kv/kvserver/gc: TestEndToEndGC failed

kv/kvserver/gc.TestEndToEndGC [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7701237?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7701237?buildTab=artifacts#/) on master @ [1a6e9f885baa124d5ff2996adb966ea15a1a9b2b](https://github.com/cockroachdb/cockroach/commits/1a6e9f885baa124d5ff2996adb966ea15a1a9b2b):


```
=== RUN   TestEndToEndGC
I221126 09:04:17.149714 1 (gostd) rand.go:193  [-] 2  random seed: 3664206823332484930
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f27dc6f41a7e04b7d5e45680c447408b/logTestEndToEndGC487535620
    test_log_scope.go:79: use -show-logs to present logs inline
    gc_int_test.go:177: range 53 stats: contains_estimates:0 last_update_nanos:1669453464698890098 intent_age:0 gc_bytes_age:0 live_bytes:0 live_count:0 key_bytes:0 key_count:0 val_bytes:0 val_count:0 intent_bytes:0 intent_count:0 separated_intent_count:0 range_key_count:0 range_key_bytes:0 range_val_count:0 range_val_bytes:0 sys_bytes:589 sys_count:7 abort_span_bytes:0 
    gc_int_test.go:178: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1607/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/gc/gc_test_/gc_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/gc/gc_int_test.go:178
        	Error:      	Should NOT be empty, but was []
        	Test:       	TestEndToEndGC
        	Messages:   	all table ranges are empty according to MVCCStats
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f27dc6f41a7e04b7d5e45680c447408b/logTestEndToEndGC487535620
--- FAIL: TestEndToEndGC (7.68s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEndToEndGC.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-21822

kv/kvserver/gc.TestEndToEndGC [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7712048?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7712048?buildTab=artifacts#/) on master @ [1a6e9f885baa124d5ff2996adb966ea15a1a9b2b](https://github.com/cockroachdb/cockroach/commits/1a6e9f885baa124d5ff2996adb966ea15a1a9b2b):


```
=== RUN   TestEndToEndGC
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f27dc6f41a7e04b7d5e45680c447408b/logTestEndToEndGC862302468
    test_log_scope.go:79: use -show-logs to present logs inline
    gc_int_test.go:177: range 53 stats: contains_estimates:0 last_update_nanos:1669538786934103719 intent_age:0 gc_bytes_age:0 live_bytes:0 live_count:0 key_bytes:0 key_count:0 val_bytes:0 val_count:0 intent_bytes:0 intent_count:0 separated_intent_count:0 range_key_count:0 range_key_bytes:0 range_val_count:0 range_val_bytes:0 sys_bytes:661 sys_count:8 abort_span_bytes:0 
    gc_int_test.go:178: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2298/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/gc/gc_test_/gc_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/gc/gc_int_test.go:178
        	Error:      	Should NOT be empty, but was []
        	Test:       	TestEndToEndGC
        	Messages:   	all table ranges are empty according to MVCCStats
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f27dc6f41a7e04b7d5e45680c447408b/logTestEndToEndGC862302468
--- FAIL: TestEndToEndGC (7.90s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEndToEndGC.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver/gc.TestEndToEndGC [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7828259?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/7828259?buildTab=artifacts#/) on master @ [859e1e1f26d79af43cce16519f9ba984baaa495c](https://github.com/cockroachdb/cockroach/commits/859e1e1f26d79af43cce16519f9ba984baaa495c):


```
=== RUN   TestEndToEndGC
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/f27dc6f41a7e04b7d5e45680c447408b/logTestEndToEndGC1005712942
    test_log_scope.go:79: use -show-logs to present logs inline
    gc_int_test.go:177: range 54 stats: contains_estimates:0 last_update_nanos:1670149071649077194 intent_age:0 gc_bytes_age:0 live_bytes:0 live_count:0 key_bytes:0 key_count:0 val_bytes:0 val_count:0 intent_bytes:0 intent_count:0 separated_intent_count:0 range_key_count:0 range_key_bytes:0 range_val_count:0 range_val_bytes:0 sys_bytes:603 sys_count:7 abort_span_bytes:0 
    gc_int_test.go:178: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/2569/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/gc/gc_test_/gc_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/gc/gc_int_test.go:178
        	Error:      	Should NOT be empty, but was []
        	Test:       	TestEndToEndGC
        	Messages:   	all table ranges are empty according to MVCCStats
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/f27dc6f41a7e04b7d5e45680c447408b/logTestEndToEndGC1005712942
--- FAIL: TestEndToEndGC (7.51s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEndToEndGC.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/gc/gc_int_test.go
+++ b/pkg/kv/kvserver/gc/gc_int_test.go
@@ -171,13 +171,27 @@ func TestEndToEndGC(t *testing.T) {
 
 	require.NotEmptyf(t, readSomeKeys(t, sqlDb), "found no keys in table")
 
-	rangeIDs := getTableRangeIDs(t, sqlDb)
-	require.NotEmpty(t, rangeIDs, "failed to query ranges belonging to table")
-
-	nonEmptyRangeIDs := findNonEmptyRanges(t, rangeIDs)
+	// Since ranges query and checking range stats are non atomic there could be
+	// a range split/merge operation caught in between. That could produce empty
+	// or incomplete results. Moreover, range info produced by ranges doesn't
+	// provide start/end keys for the range in binary form, so it is hard to make
+	// consistency check. We rely on retrying several times for simplicity.
+	const tableRangesRetry = 3
+	var tableRangeIDs, nonEmptyRangeIDs ids
+	for i := 0; i < tableRangesRetry; i++ {
+		tableRangeIDs = getTableRangeIDs(t, sqlDb)
+		if len(tableRangeIDs) == 0 {
+			continue
+		}
+		nonEmptyRangeIDs = findNonEmptyRanges(t, tableRangeIDs)
+		if len(nonEmptyRangeIDs) > 0 {
+			break
+		}
+	}
+	require.NotEmpty(t, tableRangeIDs, "failed to query ranges belonging to table")
 	require.NotEmptyf(t, nonEmptyRangeIDs, "all table ranges are empty according to MVCCStats")
 
-	deleteRangeDataWithRangeTombstone(t, rangeIDs, kvDb, sqlDb)
+	deleteRangeDataWithRangeTombstone(t, tableRangeIDs, kvDb, sqlDb)
 
 	require.Empty(t, readSomeKeys(t, sqlDb), "table still contains data after range deletion")
 

